clients/venice-thin-client/src/test/java/com/linkedin/venice/client/store/AbstractAvroStoreClientTest.java

```diff
@@ -330,4 +355,93 @@ public void testComputeReceiveNon200Response() throws ExecutionException, Interr
         storeClient.compute(Optional.empty(), Optional.empty(), 0).project("int_field").execute(keys);
     computeFuture.get();
   }
+
+  @Test
+  public void testStoreInitAsyncRetry() {
+    TransportClient mockTransportClient = new TransportClient() {
+      private final ObjectMapper OBJECT_MAPPER = ObjectMapperFactory.getInstance();
+      int retryCnt = 0;
+      int totalFailedRetryCnt = 50;
+
+      @Override
+      public CompletableFuture<TransportClientResponse> get(String requestPath, Map<String, String> headers) {
+        CompletableFuture<TransportClientResponse> result = new CompletableFuture<>();
+
+        if (requestPath.contains(TYPE_KEY_SCHEMA)) {
+          if (++retryCnt <= totalFailedRetryCnt) {
+            // Fail
+            result.completeExceptionally(new VeniceException("Fake request failure for key schema request"));
+          } else {
+            SchemaResponse keySchemaResponse = new SchemaResponse();
+            keySchemaResponse.setSchemaStr("\"string\"");
+            keySchemaResponse.setId(1);
+            try {
+              result.complete(
+                  new TransportClientResponse(
+                      -1,
+                      CompressionStrategy.NO_OP,
+                      OBJECT_MAPPER.writeValueAsBytes(keySchemaResponse)));
+            } catch (Exception e) {
+              throw new RuntimeException(e);
+            }
+          }
+          return result;
+        } else if (requestPath.contains(TYPE_STORAGE)) {
+          // Not found
+          result.complete(null);
+        } else {
+          result.completeExceptionally(new VeniceException("Fake request failure for path: " + requestPath));
+        }
+        return result;
+      }
+
+      @Override
+      public CompletableFuture<TransportClientResponse> post(
+          String requestPath,
+          Map<String, String> headers,
+          byte[] requestBody) {
+        return null;
+      }
+
+      @Override
+      public void streamPost(
+          String requestPath,
+          Map<String, String> headers,
+          byte[] requestBody,
+          TransportClientStreamingCallback callback,
+          int keyCount) {
+
+      }
+
+      @Override
+      public void close() throws IOException {
+
+      }
+    };
+    SimpleStoreClient<String, GenericRecord> storeClient = new SimpleStoreClient<>(
+        mockTransportClient,
+        "test_store_init_retry",
+        true,
+        AbstractAvroStoreClient.getDefaultDeserializationExecutor(),
+        false);
+    storeClient.setAsyncStoreInitSleepIntervalMs(1);
+    storeClient.start();
+    String testKey = "test_key";
+    try {
+      storeClient.get(testKey);
+    } catch (VeniceException e) {
+      Assert.assertTrue(e.getMessage().contains("Failed to initializing Venice Client"));
+    }
```

Comment:
Could use `Assert.expectThrows` for a more compact code.

Reply:
Sure

Reply:
Actually, `Assert.expectThrows` doesn't validate the thrown message, but I would like to make sure the right exception gets thrown.
Will leave it as it is.

Reply:
Ok. expectThrows validates exception class and returns the exception which can then be checked.

Reply:
Great, will update it.